File: src/main/java/ankol/mod/merger/merger/xml/node/XmlContainerNode.java
package ankol.mod.merger.merger.xml.node;

import lombok.Getter;
import lombok.Setter;
import lombok.ToString;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * XML container node - represents XML tags with child elements
 * Example: <skill>...</skill>
 *
 * @author Ankol
 */
@Getter
@Setter
@ToString
public class XmlContainerNode extends XmlNode {
    /**
     * Child node mapping, key is node signature, value is node object
     */
    private final Map<String, XmlNode> children = new LinkedHashMap<>();

    /**
     * Constructor
     *
     * @param signature        node signature (unique identifier)
     * @param startTokenIndex  starting token index in TokenStream
     * @param stopTokenIndex   ending token index in TokenStream
     * @param line             line number
     * @param sourceText       original source text
     */
    public XmlContainerNode(String signature, int startTokenIndex, int stopTokenIndex, int line, String sourceText) {
        super(signature.trim(), startTokenIndex, stopTokenIndex, line, sourceText);
    }

    /**
     * Add child node
     */
    public void addChild(XmlNode node) {
        children.put(node.getSignature(), node);
    }

    /**
     * Get all child nodes
     */
    public Map<String, XmlNode> getChildren() {
        return children;
    }
}

File: src/main/java/ankol/mod/merger/merger/xml/node/XmlContainerNode.kt
package ankol.mod.merger.merger.xml.node

import ankol.mod.merger.tools.logger
import org.antlr.v4.runtime.CommonTokenStream

/**
 * XML容器节点，标识可以有子节点的容器
 * 例如：<skills> ... </skills>
 * 
 * @author Ankol
 */
class XmlContainerNode(
    signature: String,
    startTokenIndex: Int,
    stopTokenIndex: Int,
    line: Int,
    tokenStream: CommonTokenStream,
    attributes: MutableMap<String, String>
) : XmlNode(signature, startTokenIndex, stopTokenIndex, line, tokenStream, attributes) {
    private val log = logger()
    val childrens: MutableMap<String, XmlNode> = LinkedHashMap()

    /**
     * 添加子节点
     */
    fun addChild(node: XmlNode) {
        if (childrens.contains(node.signature)) {
            log.debug("Repeatable siginature detected: [${node.signature}] Line: ${node.lineNumber} SourceText: ${node.sourceText}.")
        }
        childrens[node.signature] = node
    }

    override fun printTree(indent: String) {
        super.printTree(indent)
        for ((_, child) in childrens) {
            child.printTree("$indent  ")
        }
    }
}
